### ISSUE TITLE
New directive: disable=all

### ISSUE THREAD

**Issue by matthewpersico** [koalaman/shellcheck#1144]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2102, SC2171, SC2215, SC2215
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!bash
echo "this is a dummy script"
if [ "$1" = '-h' ]
then
    printf '%s\n' "$(sed '0,/^__USAGE__$/d' "$0")"
else
    echo "I worked"
fi

{
exit 0 ## In case the script does not exit properly above    
__USAGE__
myScript [-w] [--nopager] [file [file...] ]

Some description goes here.

       -w - Do not compare whitespace
--nopager - Do not pass the output through 'less'
}

```

#### Here's what shellcheck currently says:
```
Line 14	SC2102: Ranges can only match single chars (mentioned due to duplicates).
Line 14	SC2171: Found trailing ] outside test. Missing [?
Line 18	SC2215: This flag is used as a command name. Bad line break or missing [ .. ]?
Line 19	SC2215: This flag is used as a command name. Bad line break or missing [ .. ]?
```
#### Here's what I wanted or expected to see:

Because the text at the end of the script is after the exit call, the interpreter will never see it. Therefore, no checking should be done.

However, I do realize that if I added a function after the exit, it might be hard to parse.

Could we get a new directive? 

#### shellcheck disable=all

In this case I would use it as:

```sh
# shellcheck disable=all
{
exit 0 ## In case the script does not exit properly above    
...
}
```
and now I can have my usage text and shellcheck. The two braces allow me to use your existing scoping rules.

**Comment by automorphism88** [comment]
I can see this being useful in some cases, but for your script, why not put the usage message in a heredoc, and avoid the call to sed, and extra complication? You can either do something like this:

```
usage() {
    cat - <<'EOF'
Usage message goes here
EOF
}
```
or, you can go further and avoid using any external processes if you add a function to replace cat, something like the following:
```
cat_stdin() {
    local tmp
    read -r -d '' tmp && true
    # the '&& true' will avoid triggering set -e
    printf '%s' "$tmp"
}
```
in which case you'd just replace `cat - <<'EOF'` with `cat_stdin <<'EOF'` in the `usage()` function. (You can also combine them into one function if you want.) See the [bash manual](https://www.gnu.org/software/bash/manual/html_node/Redirections.html#Here-Documents) for more information on heredocs.

Also your shebang `#!bash` is invalid, unless bash is located in the current directory, since `PATH` lookups aren't done for shebangs. You should specify the absolute path, e.g. `#!/bin/bash`, or if you need the `PATH` lookup, you can use the env trick: `#!/usr/bin/env bash` which will use the first bash in your `PATH` (but has other side effects you may not want). [This Wikipedia article](https://en.wikipedia.org/wiki/Shebang_(Unix)#Portability) explains how it works.

**Comment by matthewpersico** [comment]
Thanks for your suggestions. This short reproducible is actually in a much longer set of code. The "#!bash" was just to get shellecheck.net to shut up about no shebang line; the real code has something more appropriate.

What the more elaborate code would have shown is that I am not only doing this for scripts - I am also providing the same functionality for functions. Since functions are loaded into the environment, I can't have a 'usage' function in each function script - as I source them in, they would clobber each other. And having a 'foobar_usage' function for each function foobar would clutter the environment.

Even though I am in bash, I use the ksh convention of keeping all my functions in directories that I delineate in an FPATH envvar. In this way, finding the source for any function (because I keep each function in a public interface in it's own similarly-named file) is as simple as 

    find $(echo $FPATH | tr ':' ' ') --name $func_name

I don't exactly do that, however - that `find` is buried in a personally modifed version of the `autoload` demo script that comes with the bash source code.

Anyway, the code for function `func-usage` might make things clearer:

```
## -*- sh -*-

# <Function Justification: env access>

func-usage ()
{
    ##@@ none||none||func-usage||Reads '${FUNCNAME}' or first arg for a __[EO]USAGE__ tag pair and prints everything between.
    if [ "$1" = '-h' ]
    then
        func-usage ## Nice how it works on itself, no?
    else
        ## You might pass the function name if you are in a wrapper function
        ## and the help is in the wrapped function OR if you want to call this
        ## at the command line, as a 'man' substitute.
        local funcname="$1" 
        
        ## If function name not passed in, assume the calling function.
        [ -z "$funcname" ] && funcname="${FUNCNAME[1]}"
 
        tgt=$(autoload -w "$funcname") ## A personally modified version of the
                                       ## autoload demo script that comes with
                                       ## bash. -w is "where" and it locates
                                       ## the file in which the function was
                                       ## defined (effectively a find command
                                       ## on FPATH).
 
        [ -z "$tgt" ] && return "$NOT_OK"

        ## And once we have a filename, we just call script-usage on it.
        script-usage "$tgt"
    fi
}

return 0

/dev/null<<'EOSTRINGX'
__USAGE__
func-usage - search the file defining the function for usage text and print it

The text is identified by placing the word __USAGE__ on its own line at the end
of your function script, followed it by usage text and the __EOUSAGE__
marker. Any time in your function that you want to display the text (as part of
-help processing perhaps?) just call the 'func-usage' function. It will read
the FUNCNAME array to determine the calling function (or use '$1' if you
provide it), get the corresponding file for the function, parse that file for
the __USAGE__ marker and print everything beyond it up to the __EOUSAGE__
marker.

You must ensure that your usage text will not throw an error or be executed
when the file is sourced. The easiest way to do this is with the following
format:

  #<-- is the first column for this example. The '__' variables must start in
  # the first column of each line.
  return 0 # do not use exit 0 here - function files are sourced in and an
           # exit command will exit the running process

  /dev/null<<'EOSTRING'
  __USAGE__
  Your usage text here and below...
  ...
  The last line to print.
  __EOUSAGE__
  EOSTRING

* The 'return 0' statement is used to prevent bash from reading any of the
usage text when the file is sourced in.

* The markers with the '__' are the markers used by 'func-usage' to locate the
usage text.

* The assignment of the text as a string (with the EOSTRING marker) prevents
the shell linter 'shellcheck' from complaining about syntax of text.

* Assignment to '/dev/null' prevents creating an unused variable in that
process.

__EOUSAGE_
EOSTRINGX
```

And here is `script-usage:`
```
# -*- sh -*-

# <Function Class: util>
# <Function Justification: assumes $0>

script-usage ()
{
    ##@@ none||none||script-usage||Reads '$0' or first arg for a __[EO]USAGE__ tag pair and prints everything between.
    if [ "$1" = '-h' ]
    then
        func-usage ## Can we get any more meta? :-)
    else
        ## You might pass the script name if you are in a wrapper script
        ## and the help is in the wrapped script OR if you want to call this
        ## at the command line, as a 'man' substitute.
        declare tgt="$1"
        [ -z "$tgt" ] && tgt="$0"
        [ ! -r "$tgt" ] && tgt=$(which $tgt)
        [ -z "$tgt" ] && return "$NOT_OK" ## We get an error message from the which command 
                                          ## if not found
        ## TODO: Look for marker: if __USAGE__ then printf, if __POD__ then pod2text
        printf '%s\n' "$(sed -e '0,/^__USAGE__$/d' -e '/^__EOUSAGE__/,$d' "$tgt")"

        ## If this script is a wrapper around another script, and you don't
        ## want to repeat all of the wrapped help text in the wrapper help,
        ## just put a script-usage command in your help between these chain
        ## markers and we'll execute that, sucking in that help too. We assume
        ## that the chained help comes after your help. TODO: Allow for
        ## embedded help in the above text.
        declare chain
        chain="$(sed -e '0,/^__USAGE_CHAIN__$/d' -e '/^__EOUSAGE_CHAIN__/,$d' "$tgt")"
        [ -n "$chain" ] && $chain
    fi
}

return 0

/dev/null<<'EOSTRINGX'
__USAGE__
script-usage - search the script being run for usage text and print it

The text is identified by placing the word __USAGE__ on its own line at the end
of your function script, followed it by usage text and the __EOUSAGE__
marker. Any time in your script that you want to display the text (as part of
-help processing perhaps?) just call the 'script-usage' function. It will read
the script '$0' (or use '$1' if you provide it), parse that script for the
__USAGE__ marker and print everything beyond the marker up to the __EOUSAGE
marker.

You must ensure that your usage text will not throw an error or be executed
when the script is run. The easiest way to do this is with the following
format:

  #<-- is the first column for this example. The '__' variables must start in
  # the first column of each line.
  exit 0

  /dev/null<<EOSTRING
  __USAGE__
  Your usage text here and below...
  ...
  The last line to print.
  __EOUSAGE__
  EOSTRING

* The 'exit 0' statement is used to prevent bash from reading any of the usage
text when the script is executed. ( If you don't already explicitly exit in
your script, you do now. You're welcome. :-). )

* The markers with the '__' are the markers used by 'script-usage' to locate
the usage text.

* The assignment of the text as a string (with the EOSTRING marker) prevents
the shell linter 'shellcheck' from complaining about syntax of text.

* Assignment to '/dev/null' prevents creating an unused variable in that
process.

__EOUSAGE__
EOSTRINGX
```

**Comment by automorphism88** [comment]
Interesting use of this technique of having a script use sed on itself - the main application I was aware of for doing something like this is self-extracting scripts. Thanks for sharing!

**Comment by matthewpersico** [comment]
There's other things I can do too. Being a Perl aficionado, I would prefer to do all my documentation in POD. If you look at `script-usage` you'll see the lines:

```
## TODO: Look for marker: if __USAGE__ then printf, if __POD__ then pod2text
printf '%s\n' "$(sed -e '0,/^__USAGE__$/d' -e '/^__EOUSAGE__/,$d' "$tgt")"
```

That could easily translate into (not syntax checked...):

```
if grep -q -E /^__PODUSAGE__/ "$tgt"; then pod2txt "$tgt"
elif grep -q -E /^__USAGE__/; then printf '%s\n' "$(sed -e '0,/^__USAGE__$/d' -e '/^__EOUSAGE__/,$d' "$tgt")"
else echo "No usage text found for $tgt"
fi
```
I could then more easily write man-looking pages, and I could even make the translator variable:

```
func-usage --html someFunc
```

and replace `pod2txt` with `pod2${opt_fmt}` where `fmt=html` in this case.

**Comment by harleypig** [comment]
What's wrong with

    : <<POD
    pod scattered throughout your script.
    POD
Then somewhere in your script:

    function usage() {
      pod2usage -msg "$1" "${BASH_SOURCE[0]}"
    }

**Comment by matthewpersico** [comment]
: <<POD

The colon is a new one on me.  I’ll give that a try. I may still document the script at the end instead of scattered in the script because it’s not like a Perl module with functions and methods, but I can certainly get rid of the __PODUSAGE__ marker once I convert them all to Pod.

**Comment by matthewpersico** [comment]
And I still need sed and markers when I use plain old text...

**Comment by matthewpersico** [comment]
Given that there are a billion [sic] ways in this thread to avoid the issues I have been having, no need to keep it open.

**Comment by tddschn** [comment]
Is there a way to disable all checks in version `0.7.2`?

**Comment by brother** [comment]
@tddschn something like what's proposed at #2323 maybe?

**Comment by koalaman** [comment]
In the meantime, ShellCheck started supporting ranges in disable statements so in 0.7.2 you can do `# shellcheck disable=0-9999`. `disable=all` is now a shortcut for this. Thanks!

**Comment by tddschn** [comment]
@brother Yeah, that's what I need, thank you!
